spring-cloud-kubernetes-commons/src/main/java/org/springframework/cloud/kubernetes/commons/config/ConditionalOnConfigMapPropertiesRetryDisabled.java

```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright 2013-2021 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.kubernetes.commons.config;
+
+import java.lang.annotation.Documented;
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Inherited;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+import org.springframework.boot.autoconfigure.condition.NoneNestedConditions;
+import org.springframework.context.annotation.Conditional;
+
+/**
+ * {@link Conditional @Conditional} that matches when at least one of Spring Cloud
+ * Kubernetes, Kubernetes ConfigMap property sources or Kubernetes ConfigMap property
+ * sources fail fast (thus retry) is disabled.
+ *
+ * @author Isik Erhan
+ */
+@Target({ ElementType.TYPE, ElementType.METHOD })
+@Retention(RetentionPolicy.RUNTIME)
+@Documented
+@Inherited
+@Conditional(ConditionalOnConfigMapPropertiesRetryDisabled.OnConfigMapPropertiesRetryDisabled.class)
+public @interface ConditionalOnConfigMapPropertiesRetryDisabled {
+
+	class OnConfigMapPropertiesRetryDisabled extends NoneNestedConditions {
```

Comment:
This is the first time I have used or reviewed some code with `NoneNestedConditions`, this in turn sparked my interest. I removed this part of the code below (after cloning your repo locally) : 

```
@ConditionalOnConfigMapPropertiesRetryEnabled
static class OnConfigMapPropertiesRetryEnabled {

}
```

did a `mvn clean package` on the whole project and to my surprise no test failed. :) Interesting. Took a while, but I found no way to make such a test. Ideally, such a test should be present in `KubernetesBootstrapConfigurationTests::SecretsFailFastEnabled::shouldDefineRequiredBeans`. The problem goes back to the one that we already talked before, that `@Retryable` and the need for the same interceptor name for the case when we want and not, to retry. 

On the other hand, this exercise made me go through the code and tests is much more detail and appreciate the effort you have put into this.

Reply:
This actually is about the order of the bean definitions in `KubernetesBootstrapConfiguration.RetryConfiguration`. It'd work even without annotating `configMapPropertiesRetryInterceptorNoRetry` method with `@ConditionalOnConfigMapPropertiesRetryDisabled` just beacuse of the order they're defined. If you change the bean definition order like below, some tests must start failing as you expected:

```java
@Bean("configMapPropertiesRetryInterceptor")
@ConditionalOnConfigMapPropertiesRetryDisabled
public RetryOperationsInterceptor configMapPropertiesRetryInterceptorNoRetry() {
	return RetryInterceptorBuilder.stateless().retryPolicy(new NeverRetryPolicy()).build();
}

@Bean
@ConditionalOnConfigMapPropertiesRetryEnabled
public RetryOperationsInterceptor configMapPropertiesRetryInterceptor(
		ConfigMapConfigProperties configProperties) {
	return retryOperationsInterceptor(configProperties.getRetry());
}
```

At this point, If you put the code block back that you removed, the tests must start running successfully again. With a correct ordering, `@ConditionalOnConfigMapPropertiesRetryDisabled` may seem a little bit useless, but I don't really like to rely on the order of something.

Reply:
This actually is about the order of the bean definitions in `KubernetesBootstrapConfiguration.RetryConfiguration`. It'd work even without annotating `configMapPropertiesRetryInterceptorNoRetry` method with `@ConditionalOnConfigMapPropertiesRetryDisabled` just beacuse of the order they're defined. If you change the bean definition order like below, some tests must start failing as you expected:

```java
@Bean("configMapPropertiesRetryInterceptor")
@ConditionalOnConfigMapPropertiesRetryDisabled
public RetryOperationsInterceptor configMapPropertiesRetryInterceptorNoRetry() {
	return RetryInterceptorBuilder.stateless().retryPolicy(new NeverRetryPolicy()).build();
}

@Bean
@ConditionalOnConfigMapPropertiesRetryEnabled
public RetryOperationsInterceptor configMapPropertiesRetryInterceptor(
		ConfigMapConfigProperties configProperties) {
	return retryOperationsInterceptor(configProperties.getRetry());
}
```

At this point, if you put the code block that you removed back, the tests must start running properly again. With a correct ordering, `@ConditionalOnConfigMapPropertiesRetryDisabled` may seem a little bit useless, but I don't really like to rely on the order of something. It just provides extra safety here.

Reply:
yeah, I had to take a closer look today and this is the same conclusion I came to. thank you.